### PR TITLE
Add tests to configurable query backend PR

### PR DESCRIPTION
**Pull request by uhbif19**
For https://github.com/Plutonomicon/cardano-transaction-lib/pull/1260

### Pre-review checklist

- [x] All code has been formatted using our config (`make format`)
- [x] Any new API features or modification of existing behavior is covered as defined in our [test plan](https://github.com/Plutonomicon/cardano-transaction-lib/blob/develop/doc/test-plan.md)
- [x] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

File: test/Plutip/Contract.purs [hunk 1]
```diff
@@ -784,6 +795,80 @@ suite = do
...
+        withKeyWallet alice do
+          validator1 <- AlwaysSucceeds.alwaysSucceedsScript
+          validator2 <- alwaysSucceedsScriptV2
+          let
+            validatorRef1 = PlutusScriptRef $ unwrap validator1
+            validatorRef2 = PlutusScriptRef $ unwrap validator2
+            useScriptAndGetByHash validator vhash = do
+              txId <- AlwaysSucceeds.payToAlwaysSucceeds vhash
+              awaitTxConfirmed txId
+              -- For Kupo (used inside) to see script, its utxo must be spent
```
Comment (jy14898) [inline test/Plutip/Contract.purs:813]: This isn't entirely true, Kupo will also index script references in outputs. For eg `mustPayToScriptWithScriptRef` is enough to make kupo aware in 1 transaction. That being said, what we have now is probably easier

File: test/Plutip/Contract.purs [hunk 2]
```diff
@@ -784,6 +795,80 @@ suite = do
...
+
+          let
+            unpackResult' :: Partial => _
+            unpackResult' (Right (Just x)) = x
+            unpackResult' _ = crash "test fail: No script found by hash"
+            unpackResult x = unsafePartial $ unpackResult' x
+
+          -- Testing getScriptByHash
+          (unpackResult result1) `shouldEqual` validatorRef1
+          (unpackResult result2) `shouldEqual` validatorRef2
```
Comment (jy14898) [inline test/Plutip/Contract.purs:828]: I'm not a fan of this, I prefer adding the constructors to the target

```suggestion
          -- Testing getScriptByHash
          result1 `shouldEqual` (Right (Just validatorRef1))
          result2 `shouldEqual` (Right (Just validatorRef2))
```

File: test/Plutip/Contract.purs [hunk 3]
```diff
@@ -123,6 +127,11 @@ import Ctl.Internal.Plutus.Types.Value (lovelaceValueOf)
 import Ctl.Internal.Scripts (nativeScriptHashEnterpriseAddress)
 import Ctl.Internal.Test.TestPlanM (TestPlanM)
 import Ctl.Internal.Types.Interval (getSlotLength)
+import Ctl.Internal.Types.TransactionMetadata
```
Comment (jy14898) [inline test/Plutip/Contract.purs:130]: I know we have loads of internal imports already, but lets try and import from the public interface where possible, as it adds another assurance to our API

File: test/Plutip/Contract.purs [hunk 4]
```diff
@@ -784,6 +795,80 @@ suite = do
...
+
+      withWallets distribution \alice -> do
+        withKeyWallet alice do
+          tn <- mkTokenName "Token name"
+          mp /\ _ <- mkCurrencySymbol alwaysMintsPolicy
+          let
+            constraints :: Constraints.TxConstraints Void Void
+            constraints = mconcat
+              [ Constraints.mustMintCurrency (mintingPolicyHash mp) tn one
+              ]
```
Comment (jy14898) [inline test/Plutip/Contract.purs:853]:
```suggestion
            constraints =
              Constraints.mustMintCurrency (mintingPolicyHash mp) tn one
```

**Review by jy14898 — CHANGES_REQUESTED** [review]

File: test/Plutip/Contract.purs (hunk 1 above)
Comment (uhbif19) [reply test/Plutip/Contract.purs:813]: Thanks. I did not found mention of what you say in Kupo API docs. It says:

> However, a periodic garbage collector will automatically remove scripts that aren't associated to any matched output

https://cardanosolutions.github.io/kupo/#tag/Scripts

Comment (jy14898) [reply test/Plutip/Contract.purs]: > By default, the engine indexes every script present in transaction witnesses and inline scripts

Only a transaction which spends a script output will include the script in the witness set, but we also have inline scripts in utxos which will get indexed by Kupo.

File: test/Plutip/Contract.purs [hunk 5]
```diff
@@ -784,6 +795,73 @@ suite = do
...
+
+          -- Testing getScriptsByHashes
+          let
+            scriptHash1 = unwrap (validatorHash validator1)
+            scriptHash2 = unwrap (validatorHash validator2)
+          results <- getScriptsByHashes [ scriptHash1, scriptHash2 ]
+          (map unpackResult results) `shouldEqual` Map.fromFoldable
+            [ (scriptHash1 /\ validatorRef1)
+            , (scriptHash2 /\ validatorRef2)
+            ]
```
Comment (jy14898) [inline test/Plutip/Contract.purs:832]: I meant both uses of `unpackResult` in my previous change request, my previous change on its own does not compile, it required follow up changes

```suggestion
          results <- getScriptsByHashes [ scriptHash1, scriptHash2 ]
          results `shouldEqual` Map.fromFoldable
            [ (scriptHash1 /\ (Right (Just validatorRef1)))
            , (scriptHash2 /\ (Right (Just validatorRef2)))
            ]
```

File: test/Plutip/Contract.purs [hunk 6]
```diff
@@ -784,6 +795,73 @@ suite = do
...
+          tn <- mkTokenName "Token name"
+          mp /\ _ <- mkCurrencySymbol alwaysMintsPolicy
+          let
+            constraints :: Constraints.TxConstraints Void Void
+            constraints =
+              Constraints.mustMintCurrency (mintingPolicyHash mp) tn one
+
+            lookups :: Lookups.ScriptLookups Void
+            lookups =
+              Lookups.mintingPolicy mp
```
Comment (jy14898) [inline test/Plutip/Contract.purs:850]: I think it's sufficient to have `constraints` and `lookups` be `mempty`? I assume you took the following test as a base for the example

File: test/Plutip/Contract.purs [hunk 7]
```diff
@@ -784,6 +795,73 @@ suite = do
...
+          ubTx <- liftedE $ Lookups.mkUnbalancedTx lookups constraints
+          ubTx' <- setGeneralTxMetadata ubTx givenMetadata
+          bsTx <- signTransaction =<< liftedE (balanceTx ubTx')
+          txId <- submit bsTx
+          awaitTxConfirmed txId
+
+          mMetadata <- getTxMetadata txId
+          let metadata = unsafePartial $ fromJust mMetadata
+
+          metadata `shouldEqual` givenMetadata
```
Comment (jy14898) [inline test/Plutip/Contract.purs:863]:
```suggestion
          mMetadata `shouldEqual` (Just givenMetadata)
```

**Review by jy14898 — CHANGES_REQUESTED** [review]

File: test/Plutip/Contract.purs (hunk 5 above)
Comment (uhbif19) [reply test/Plutip/Contract.purs:832]: I understand, was about to fix that, thanks.

**Review by jy14898 — APPROVED** [review]
LGTM

**Comment by uhbif19** [comment]
Should probably wait for merging this and rebase on it: https://github.com/Plutonomicon/cardano-transaction-lib/pull/1371

**Pull request merged**